File: src/test/java/com/koddy/server/coffeechat/domain/model/link/MeetingLinkProviderTest.java
package com.koddy.server.coffeechat.domain.model.link;

import com.koddy.server.coffeechat.exception.CoffeeChatException;
import com.koddy.server.common.UnitTest;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static com.koddy.server.coffeechat.domain.model.link.MeetingLinkProvider.GOOGLE;
import static com.koddy.server.coffeechat.domain.model.link.MeetingLinkProvider.ZOOM;
import static com.koddy.server.coffeechat.exception.CoffeeChatExceptionCode.INVALID_MEETING_LINK_PROVIDER;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.jupiter.api.Assertions.assertAll;

@DisplayName("CoffeeChat/Link -> MeetingLinkProvider 테스트")
class MeetingLinkProviderTest extends UnitTest {
    @Test
    @DisplayName("제공하지 않는 Provider에 대해서 MeetingLinkProvider를 가져오려고 하면 예외가 발생한다")
    void throwExceptionByInvalidMeetingLinkProvider() {
        assertThatThrownBy(() -> MeetingLinkProvider.from("anonymous"))
                .isInstanceOf(CoffeeChatException.class)
                .hasMessage(INVALID_MEETING_LINK_PROVIDER.getMessage());
    }

    @Test
    @DisplayName("주어진 Provider에 따른 MeetingLinkProvider를 가져온다")
    void getSpecificMeetingLinkProvider() {
        assertAll(
                () -> assertThat(MeetingLinkProvider.from("zoom")).isEqualTo(ZOOM),
                () -> assertThat(MeetingLinkProvider.from("google")).isEqualTo(GOOGLE)
        );
    }
}

File: src/test/kotlin/com/koddy/server/coffeechat/domain/model/link/MeetingLinkProviderTest.kt
package com.koddy.server.coffeechat.domain.model.link

import com.koddy.server.coffeechat.exception.CoffeeChatException
import com.koddy.server.coffeechat.exception.CoffeeChatExceptionCode.INVALID_MEETING_LINK_PROVIDER
import com.koddy.server.common.UnitTestKt
import io.kotest.assertions.assertSoftly
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.annotation.DisplayName
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.throwable.shouldHaveMessage

@UnitTestKt
@DisplayName("CoffeeChat/Link -> MeetingLinkProvider 테스트")
internal class MeetingLinkProviderTest : DescribeSpec({
    describe("MeetingLinkProvider's from") {
        context("제공하지 않는 Provider는") {
            it("INVALID_MEETING_LINK_PROVIDER 예외가 발생한다") {
                shouldThrow<CoffeeChatException> {
                    MeetingLinkProvider.from("anonymous")
                } shouldHaveMessage INVALID_MEETING_LINK_PROVIDER.message
            }
        }

        context("제공하는 Provider는") {
            it("MeetingLinkProvider로 변환해서 응답한다") {
                assertSoftly {
                    MeetingLinkProvider.from("zoom") shouldBe MeetingLinkProvider.ZOOM
                    MeetingLinkProvider.from("google") shouldBe MeetingLinkProvider.GOOGLE
                }
            }
        }
    }
})
